### PR TITLE
[dev] [DeepSeek-v4] Part 1: Hybrid Attention with CSA and HCA

### PR DESCRIPTION
**Pull request by hxbai**
# What does this PR do ?

We will create several PRs to functionally support DeepSeek-v4 training. This is the first one.

Add DeepSeek-v4 Hybrid Attention with CSA and HCA.

We reuse some arguments from DSA to avoid duplication.

DeepSeek-V4 training has several phases, and here are the setting examples (use DSv4-Flash size):

- Dense mode:
    - `--experimental-attention-variant dsv4_hybrid`
    - `--csa-window-size 128`
    - `--csa-compress-ratios ([0,0,4]+[128,4]*20+[0])`
    - `--csa-compress-rotary-base: 40000`
    - `--dsa-indexer-n-heads 64`
    - `--dsa-indexer-head-dim 128`
    - `--dsa-indexer-topk 512`
    - **`--csa-dense-mode`**
- Sparse warmup mode:
    - `--experimental-attention-variant dsv4_hybrid`
    - `--csa-window-size 128`
    - `--csa-compress-ratios ([0,0,4]+[128,4]*20+[0])`
    - `--csa-compress-rotary-base: 40000`
    - `--dsa-indexer-n-heads 64`
    - `--dsa-indexer-head-dim 128`
    - `--dsa-indexer-topk 512`
- Sparse mode:
    - `--experimental-attention-variant dsv4_hybrid`
    - `--csa-window-size 128`
    - `--csa-compress-ratios ([0,0,4]+[128,4]*20+[0])`
    - `--csa-compress-rotary-base: 40000`
    - `--dsa-indexer-n-heads 64`
    - `--dsa-indexer-head-dim 128`
    - `--dsa-indexer-topk 512`
    - **`--dsa-indexer-use-sparse-loss`**

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: tests/unit_tests/transformer/experimental_attention_variant/test_attention_variant_csa.py [hunk 1]
```diff
@@ -0,0 +1,864 @@
...
+from megatron.core.tensor_parallel.random import model_parallel_cuda_manual_seed
+from megatron.core.transformer.enums import AttnMaskType
+from megatron.core.transformer.experimental_attention_variant.csa import (
+    CompressedSparseAttention,
+    CompressedSparseAttentionSubmodules,
+    Compressor,
+    CompressorSubmodules,
+    CSAIndexer,
+    CSAIndexerSubmodules,
+    get_compress_topk_idxs,
```
[the remark this answers is not in the document]
Comment (hxbai) [reply tests/unit_tests/transformer/experimental_attention_variant/test_attention_variant_csa.py:18]: fixed

File: tests/unit_tests/transformer/experimental_attention_variant/test_dsv4_hybrid_attention.py [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+        hidden = torch.randn(
+            seq_len, batch_size, self.config.hidden_size, dtype=torch.bfloat16
+        ).cuda()
+
+        with torch.no_grad():
+            output, bias = attn(hidden_states=hidden, attention_mask=None)
+
+        assert output.shape == (seq_len, batch_size, self.config.hidden_size)
+        assert not torch.isnan(output).any()
+
```
[the remark this answers is not in the document]
Comment (hxbai) [reply tests/unit_tests/transformer/experimental_attention_variant/test_dsv4_hybrid_attention.py:301]: fixed

File: megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py [hunk 3]
```diff
@@ -0,0 +1,887 @@
...
+        group_proj_out_size = self.config.o_groups * self.config.o_lora_rank
+
+        _linear_o_group_proj = torch.empty(
+                group_proj_out_size,
+                group_proj_in_size,
+                device=torch.cuda.current_device(),
+                dtype=self.config.params_dtype,
+            )
+        self.config.init_method(_linear_o_group_proj)
+        self.linear_o_group_proj = torch.nn.Parameter(_linear_o_group_proj)
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py:178]: added assertion that only TP = 1 is supported

File: megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py [hunk 4]
```diff
@@ -0,0 +1,887 @@
...
+            skip_bias_add=False,
+            is_expert=False,
+            tp_comm_buffer_name='kv_up_proj',
+            tp_group=pg_collection.tp,
+        )
+        self.kv_layernorm = submodules.kv_layernorm(
+            hidden_size=self.config.v_head_dim,
+            config=self.config,
+            eps=self.config.layernorm_epsilon,
+        )
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py]: added assertion that only TP = 1 is supported

File: megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py [hunk 5]
```diff
@@ -0,0 +1,887 @@
...
+
+        # Concatenate back and reshape to original shape
+        weight_kv_updated = torch.cat([weight_k, weight_v], dim=1)
+        weight_kv_updated = weight_kv_updated.view(
+            self.num_attention_heads_per_partition
+            * (self.config.qk_head_dim + self.config.v_head_dim),
+            -1,
+        )
+
+        return weight_kv_updated
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py]: removed qk_clip support becuase DSv4 doesn't need it

File: megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py [hunk 6]
```diff
@@ -0,0 +1,887 @@
...
+        if self.linear_kv_down_proj is not None:
+            self.linear_kv_down_proj.backward_dw()
+
+    def _backward_q_proj(self):
+        """Computes weight gradients of Q projection layers"""
+        if self.config.q_lora_rank is None:
+            self.linear_q_proj.backward_dw()
+        else:
+            self.linear_q_down_proj.backward_dw()
+            self.linear_q_up_proj.backward_dw()
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py]: fixed

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 7]
```diff
@@ -0,0 +1,791 @@
...
+    kv_t = kv_full.permute(1, 0, 2)
+
+    safe_indices = topk_indices.clamp(min=0).long()  # [b, sq, topk]
+    safe_indices_exp = safe_indices.unsqueeze(-1).expand(-1, -1, -1, hn)  # [b, sq, topk, hn]
+    # [b, n_kv, hn] -> [b, 1, n_kv, hn] -> gather -> [b, sq, topk, hn]
+    kv_gathered = torch.gather(
+        kv_t.unsqueeze(1).expand(-1, sq, -1, -1), dim=2, index=safe_indices_exp
+    )
+
+    # --- Attention scores ---
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py:200]: The unfused version is just for reference and testing. It cannot be used in real scenario due to perf/memory overhead. So, I'll leave it as this.

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 8]
```diff
@@ -0,0 +1,791 @@
...
+    # query: [sq, b, np, hn] -> [b, np, sq, hn]
+    q = query.permute(1, 2, 0, 3).float()
+    kv_g = kv_gathered.float()  # [b, sq, topk, hn]
+
+    # [b, np, sq, topk]
+    scores = torch.einsum("bnsh,bskh->bnsk", q, kv_g) * softmax_scale
+
+    # Mask invalid
+    invalid_mask = (topk_indices < 0).unsqueeze(1)  # [b, 1, sq, topk]
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py:209]: The unfused version is just for reference and testing. It cannot be used in real scenario due to perf/memory overhead. So, I'll leave it as this.

File: megatron/core/models/common/embeddings/rope_utils.py [hunk 9]
```diff
@@ -132,8 +140,17 @@ def _apply_rotary_pos_emb_bshd(
     # second part is sine component, need to change signs with _rotate_half method
     cos_ = (torch.cos(freqs) * mscale).to(t.dtype)
     sin_ = (torch.sin(freqs) * mscale).to(t.dtype)
+    if inverse:
+        sin_ = -sin_
 
     t = (t * cos_) + (_rotate_half(t, rotary_interleaved) * sin_)
+
+    # Fallback to original permutation
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/models/common/embeddings/rope_utils.py:149]: added a comment

File: megatron/core/transformer/transformer_config.py [hunk 10]
```diff
@@ -2696,6 +2730,21 @@ def __post_init__(self):
...
+            assert (
+                not self.mla_down_proj_fusion
+            ), "MLA down projection fusion must be disabled for DSv4 hybrid mode."
+            log_single_rank(
+                logger,
+                logging.WARNING,
+                f"DSv4 hybrid mode is enabled, deriving qk_head_dim and kv_lora_rank from "
+                f"v_head_dim and qk_pos_emb_head_dim"
+            )
+            derived = self.v_head_dim - self.qk_pos_emb_head_dim
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/transformer_config.py:2811]: updated the doctring

File: megatron/training/arguments.py [hunk 11]
```diff
@@ -4610,6 +4638,17 @@ def _add_experimental_attention_variant_args(parser):
...
+        type=compress_ratios_type,
+        default=None,
+        help='Per-layer compress ratios for compressed sparse attention. '
+            'Accepts a string containing a Python list expression, e.g.: '
+            '"[0,0,4,128,4,128]" or "([0]+[4,128]*2)*3". '
+            'Each value is the compression ratio for the corresponding '
+            'transformer layer (valid values: 0, 4, 128). '
+            'The list length must equal num_layers.'
+    )
     return parser
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/training/arguments.py:4656]: configs will be automatically exported to args

File: megatron/core/transformer/experimental_attention_variant/csa.py [hunk 12]
```diff
@@ -0,0 +1,791 @@
...
+    rotate_activation,
+)
+from megatron.core.transformer.module import MegatronModule
+from megatron.core.transformer.spec_utils import ModuleSpec, build_module
+from megatron.core.transformer.transformer_config import TransformerConfig
+from megatron.core.utils import get_pg_size, nvtx_range_pop, nvtx_range_push
+
+
+# ---------------------------------------------------------------------------
+# Helper functions for index computation
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/csa.py:29]: Leave it for now. Will consider it when we add THD support.

File: megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py [hunk 13]
```diff
@@ -0,0 +1,887 @@
...
+            core_attn_out = self._checkpointed_attention_forward(
+                query, key, value, attention_mask, packed_seq_params=packed_seq_params
+            )
+        else:
+            extra_kwargs = {}
+            if self.config.experimental_attention_variant in ("dsa", "dsv4_hybrid"):
+                # For dsa we need to pass in the original hidden states and the compressed
+                # query representation.
+                extra_kwargs["x"] = hidden_states
+                extra_kwargs["qr"] = q_compressed
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py]: added an assertion

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4458/commits/6f5b92f4676544a77d8809ef60c2f840e92b28e5

**Comment by chenhongyu2048** [comment]
Hello @hxbai I'm also very interested in implementing sparse attention, and I'd like to know if CP/SP support will be added in the future?

**Comment by hxbai** [comment]
> Hello @hxbai I'm also very interested in implementing sparse attention, and I'd like to know if CP/SP support will be added in the future?

Yes, that's in the plan

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4458/commits/29048b0a5470e9de794ba28bc3ffdc3aa83e071e

**Review by yaox12 — APPROVED** [review]
LGTM.

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/4458/commits/8470a3b67fbae46e7e1040fa379ed74ef0634390
